src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/WrapWithTag/WrapWithTagEndpoint.cs

```diff
@@ -0,0 +1,107 @@
+using Microsoft.AspNetCore.Razor.LanguageServer.Extensions;
+using Microsoft.AspNetCore.Razor.LanguageServer.ProjectSystem;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.WrapWithTag
+{
+    internal class WrapWithTagEndpoint : IWrapWithTagHandler
+    {
+        private readonly ClientNotifierServiceBase _languageServer;
+        private readonly ProjectSnapshotManagerDispatcher _projectSnapshotManagerDispatcher;
+        private readonly DocumentResolver _documentResolver;
+        private readonly RazorDocumentMappingService _razorDocumentMappingService;
+        private readonly ILogger _logger;
+
+        public WrapWithTagEndpoint(
+            ClientNotifierServiceBase languageServer,
+            ProjectSnapshotManagerDispatcher projectSnapshotManagerDispatcher,
+            DocumentResolver documentResolver,
+            RazorDocumentMappingService razorDocumentMappingService,
+            ILoggerFactory loggerFactory)
+        {
+            if (languageServer is null)
+            {
+                throw new ArgumentNullException(nameof(languageServer));
+            }
+
+            if (projectSnapshotManagerDispatcher is null)
+            {
+                throw new ArgumentNullException(nameof(projectSnapshotManagerDispatcher));
+            }
+
+            if (documentResolver is null)
+            {
+                throw new ArgumentNullException(nameof(documentResolver));
+            }
+
+            if (razorDocumentMappingService is null)
+            {
+                throw new ArgumentNullException(nameof(razorDocumentMappingService));
+            }
+
+            if (loggerFactory is null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _languageServer = languageServer;
+            _projectSnapshotManagerDispatcher = projectSnapshotManagerDispatcher;
+            _documentResolver = documentResolver;
+            _razorDocumentMappingService = razorDocumentMappingService;
+            _logger = loggerFactory.CreateLogger<WrapWithTagEndpoint>();
+        }
+
+        public async Task<WrapWithTagResponse?> Handle(WrapWithTagParams request, CancellationToken cancellationToken)
+        {
+            var documentSnapshot = await _projectSnapshotManagerDispatcher.RunOnDispatcherThreadAsync(() =>
+            {
+                _documentResolver.TryResolveDocument(request.TextDocument.Uri.GetAbsoluteOrUNCPath(), out var documentSnapshot);
+
+                return documentSnapshot;
+            }, cancellationToken).ConfigureAwait(false);
+
+            if (documentSnapshot is null || cancellationToken.IsCancellationRequested)
+            {
+                _logger.LogWarning($"Failed to find document {request.TextDocument.Uri}.");
+                return null;
+            }
+
+            var codeDocument = await documentSnapshot.GetGeneratedOutputAsync();
```

Comment:
```suggestion
            var codeDocument = await documentSnapshot.GetGeneratedOutputAsync().ConfigureAwait(false);
```

Should probably sprinkle this later in the file too

Reply:
I keep forgetting this... seems I've gotten too used to having the anaylzer :P